Added odm:generate:persistent-collections CLI command

Bilge opened the pull request:
PHP 5.5 support is dropped because ODM also dropped support for that version and persistent collection generation is not available for that version either.

gianarb commented:
Sorry @Bilge can you rebase please?

Thanks a lot!

webimpress commented:
@gianarb I think this PR can be closed. All important changes from this PR are already on dev-master IMO. I'm not sure about changing `$this->` to `self::` in tests and adding new `bin/cli` file.

gianarb commented:
Thanks @webimpress we have not the subject of this PR if I understood well.
persistent-collections is not a command that we support right know.. or am I wrong?

webimpress commented:
@gianarb I'm not sure if we support this command (I would bet we do), so probably it could be added, but commit
https://github.com/doctrine/DoctrineMongoODMModule/pull/165/commits/1ff3cc6259d9497a83fa6fbc0900d1bb3f55a3a8
definitely contains too much changes and make this PR unclear.

IMO we shouldn't add file `bin/cli` - unused and then removing it could be BC Break.

PHP 5.5 support is already dropped.

gianarb commented:
Yes it's too big I prefer a PR with just the new command.

@Bilge can you address this?
Thanks a lot

Bilge commented:
@gianarb No. The code cleanup is good.

gianarb commented on tests/DoctrineMongoODMModuleTest/Doctrine/CliTest.php:
```diff
@@ -26,7 +34,7 @@
  * @author  Adam Homsi <adam.homsi@gmail.com>
  */
-class CliTest extends PHPUnit_Framework_TestCase
+final class CliTest extends \PHPUnit_Framework_TestCase
```
I can not merge this. We don't know how people are using CliTest, can you rollback to `class`?

Bilge replied:
@gianarb  That is completely false. This entire namespace is only present in `autoload-dev`, ergo it is not even accessible to users. They cannot be using this.

alcaeus replied:
@gianarb technically, this is a BC break, but it's acceptable since this library hasn't reached 1.0.0 yet. If you're going to merge this in, please add a BC break notice. People shouldn't be using this class anyways, but better play it safe.

Ocramius replied:
@alcaeus this NS is not public, nor is shipped with the package (it's [dev-only](https://github.com/doctrine/DoctrineMongoODMModule/blob/6f7f00f778fa4a03d28831c019992a32729f9cc1/composer.json#L69-L73), as @Bilge said)

alcaeus replied:
It's not autoloaded. It's still present as a file.

Bilge replied:
@alcaeus Are you saying that it is feasible and acceptable for someone to manually autoload test files and extend them, and therefore we must consider marking such test files as `final` a BC-break, but the only reason we don't is because the version number is less than one?

gianarb commented on src/DoctrineMongoODMModule/Module.php:
```diff
@@ -42,8 +41,7 @@
  */
-class Module implements
-    BootstrapListenerInterface,
+final class Module implements
```
Can you rollback to `class`? We don't know how people are using `Module` right know. I don't think that someone is extending this class but we can not know

Bilge replied:
I could revert this but there is no use-case for extending the `Module` class, and doing so would be incorrect. I understand that from a purist's point-of-view this would be a BC break but in practice it is not.

gianarb replied:
@Bilge I understand your point, but in practice we don't know how people are doing and in theory this is a bc break. I can not merge it

gianarb commented:
@Bilge you know how it works. Little PR are usually better that big. This is not too big you are right but you usually create PRs that covers more stuffs and @webimpress is doing a good work to review and address PRs.

I leave some comments for you and for the future please, one PR on problem. 
Change `array` to `[]` or `namespace` seems stupid but it's out of that your PR implements.

Bilge commented:
@gianarb As per your request, I have resolved the conflicts. I added a merge commit instead of rebasing so you would have better visibility of the changes. I also removed the `final` keyword from `Module` as requested.

webimpress commented on bin/cli:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env php
```
IMO we don't need this file.
1. It is not used so it shouldn't be available to users
2. It is for testing purposes, so if it is really needed it should be moved into `tests` directory

webimpress commented on tests/DoctrineMongoODMModuleTest/Doctrine/CliTest.php:
```diff
@@ -83,37 +86,17 @@ public function testValidHelpers()
+        self::assertInstanceOf(QueryCommand::class, $this->cli->get('odm:query'));
+        self::assertInstanceOf(CreateCommand::class, $this->cli->get('odm:schema:create'));
+        self::assertInstanceOf(UpdateCommand::class, $this->cli->get('odm:schema:update'));
+        self::assertInstanceOf(DropCommand::class, $this->cli->get('odm:schema:drop'));
```
Why `$this->` is changed to `self::`? I know it is also right, but I'd go with `$this->` because:
1. It is like that in the whole library - so to keep consistency
2. Even in PHPUnit documentation all examples are using `$this->`
3. PHPStorm bring back this functionality in latest release for PHPUnit usages

Bilge commented:
@webimpress I have asked you to stop commenting on my PRs. I have not asked you to stop commenting on all PRs or to stop contributing to this project, even though I would do both if it were my prerogative, I have simply asked you to stop commenting on my PRs.

Ocramius commented:
@Bilge if you don't want to read someone's comments, just feel free to ignore them, but you can't force people out of contributing. I'm locking the discussion. @gianarb, feel free to unlock when you want to work further on this.

alcaeus commented:
@Bilge Since you're adding support for custom collections in #166, could you add the generator command to that PR as well? Please leave out any refactorings and move them to a new PR if you want to get them merged. Thanks!
